core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java

```diff
@@ -1250,6 +1251,9 @@ private static Type unwrapOptional(Type parameterizedType) {
             if (type instanceof ParameterizedType) {
                 return (Class<?>) ((ParameterizedType) type).getRawType();
             }
+            if (type instanceof WildcardType) {
+                return (Class<?>) ((WildcardType) type).getUpperBounds()[0];
+            }
```

Comment:
* Check the length of `getUpperBounds()` before accessing the first element?
* Could we also consider a generic `TypeVariable` such as `<T extends Foo> foo(@Param List<T> foo)`?

Reply:
> Check the length of getUpperBounds() before accessing the first element?

👍 `IllegalArgumentException` is better than `IndexOutOfBoundException`.

 > Could we also consider a generic TypeVariable such as `<T extends Foo> foo(@Param List<T> foo)`?

I added resolving `TypeVariable`. But what do I do when bounds are multiple such as `<T extends Foo | Comparable> foo(@Param List<T> foo)`? Retrofit had the same thought and resolved to `Object`: https://github.com/square/retrofit/blob/963af06c02a8ae8193037ccaa597b5091ba95e06/retrofit/src/main/java/retrofit2/Utils.java#L88-L92

Reply:
> But what do I do when bounds are multiple such as <T extends Foo | Comparable> foo(@Param List<T> foo)

It makes sense to support a single bound.